File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let maxProfit = 0;
    let currMin = prices[0];
    for(let i = 1; i < prices.length; i++) {
        //look backwards by 1!
        //if smaller, check against current min and overwrite if smaller
        if (prices[i - 1] > prices[i]) {
            currMin = Math.min(currMin, prices[i]);
        }
        //if larger, run math to calculate profit and overwrite if larger
        if (prices[i - 1] < prices[i]){
            let tempProfit = prices[i] - currMin;
            maxProfit = Math.max(maxProfit, tempProfit);
        }
    }
    return maxProfit;
};

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int maxProfit = 0;
        int currMin = prices[0];
        int len = prices.size();
        for(int i = 1; i < len; ++i) {
            //look backwards by 1!
            //if smaller, check against current min and overwrite if smaller
            if (prices[i - 1] > prices[i]) {
                currMin = min(currMin, prices[i]);
            }
            //if larger, run math to calculate profit and overwrite if larger
            if (prices[i - 1] < prices[i]){
                int tempProfit = prices[i] - currMin;
                maxProfit = max(maxProfit, tempProfit);
            }
        }
        return maxProfit;
    }
};